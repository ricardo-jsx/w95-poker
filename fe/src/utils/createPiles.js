export default function (cards = [], piles = 0) {
  const filledPiles = [];
  let currentCardPos = 0;

  for (let pile = 0; pile < piles; pile++) {
    for (let row = 0; row <= pile; row++) {
      if (!filledPiles[pile]) filledPiles[pile] = [];

      filledPiles[pile][row] = { ...cards[currentCardPos++] };
    }

    //Last card of pile needs to be facing up
    filledPiles[pile][filledPiles[pile].length - 1].facingUp = true;
  }

  return { piles: filledPiles, length: currentCardPos };
}
