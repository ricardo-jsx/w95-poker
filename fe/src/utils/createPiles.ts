import { Card } from 'resources/cards/cards.types';

export default function (cards: Array<Card>, piles = 7): Response {
  const filledPiles: Piles = [];
  let currentCardPos = 0;

  for (let pile = 0; pile < piles; pile++) {
    for (let row = 0; row <= pile; row++) {
      if (!filledPiles[pile]) filledPiles[pile] = [];

      filledPiles[pile][row] = { ...cards[currentCardPos++] };
    }

    //Last card of pile needs to be facing up
    filledPiles[pile][filledPiles[pile].length - 1].facingUp = true;
  }

  return { piles: filledPiles, length: currentCardPos };
}

type Piles = Array<Array<SolitaireCard>>;

interface SolitaireCard extends Card {
  facingUp?: boolean;
}

interface Response {
  piles: Piles;
  length: number;
}
