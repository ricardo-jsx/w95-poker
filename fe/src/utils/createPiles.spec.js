import createPiles from './createPiles';

const cards = [{ value: 'AH' }, { value: '2H' }, { value: '3H' }, { value: '4H' }, { value: '5H' }, { value: '6H' }];

describe('utils/createPiles', () => {
  it('should create 3 piles with 6 cards', () => {
    const { piles, length } = createPiles(cards, 3);

    expect(piles).toMatchSnapshot();
    expect(length).toBe(6);
  });
});
