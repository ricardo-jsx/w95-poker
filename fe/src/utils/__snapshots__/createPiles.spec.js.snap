// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/createPiles should create 3 piles with 6 cards 1`] = `
Array [
  Array [
    Object {
      "facingUp": true,
      "value": "AH",
    },
  ],
  Array [
    Object {
      "value": "2H",
    },
    Object {
      "facingUp": true,
      "value": "3H",
    },
  ],
  Array [
    Object {
      "value": "4H",
    },
    Object {
      "value": "5H",
    },
    Object {
      "facingUp": true,
      "value": "6H",
    },
  ],
]
`;
